xds/src/main/java/io/grpc/xds/internal/sds/ServerWrapperForXds.java

```diff
@@ -77,27 +132,113 @@ public Server start() throws IOException {
       throw new RuntimeException(ex);
     }
     delegate.start();
+    currentServingState = ServingState.STARTED;
     xdsServingStatusListener.onServing();
     return this;
   }
 
-  private Future<EnvoyServerProtoData.DownstreamTlsContext> addServerWatcher() {
-    final SettableFuture<EnvoyServerProtoData.DownstreamTlsContext> settableFuture =
-        SettableFuture.create();
+  @VisibleForTesting public ServingState getCurrentServingState() {
+    return currentServingState;
+  }
+
+  private SettableFuture<Throwable> addServerWatcher() {
+    final SettableFuture<Throwable> future = SettableFuture.create();
     serverWatcher =
         new XdsClientWrapperForServerSds.ServerWatcher() {
           @Override
-          public void onError(Throwable throwable) {
-            xdsServingStatusListener.onNotServing(throwable);
+          public void onError(Throwable throwable, boolean isAbsent) {
+            if (currentServingState == ServingState.STARTING) {
+              // during start
+              if (isPermanentErrorFromXds(throwable)) {
+                currentServingState = ServingState.SHUTDOWN;
+                future.set(throwable);
+                return;
+              }
+              xdsServingStatusListener.onNotServing(throwable);
+            } else {
+              // after start
+              if (isAbsent) {
+                xdsServingStatusListener.onNotServing(throwable);
+                if (currentServingState == ServingState.STARTED) {
+                  // shutdown the server
+                  delegate.shutdown();  // let existing calls finish on delegate
+                  currentServingState = ServingState.NOT_SERVING;
+                  delegate = null;
+                }
+              }
+            }
           }
 
           @Override
-          public void onSuccess() {
-            settableFuture.set(null);
+          public void onListenerUpdate() {
+            if (currentServingState == ServingState.STARTING) {
+              // during start()
+              future.set(null);
+            } else if (currentServingState == ServingState.NOT_SERVING) {
+              // coming out of NOT_SERVING
+              currentServingState = ServingState.ENTER_SERVING;
+              startRetryTask = new StartRetryTask();
+              startRetryTask.createTask(0L);
+            }
           }
         };
     xdsClientWrapperForServerSds.addServerWatcher(serverWatcher);
-    return settableFuture;
+    return future;
+  }
+
+  private final class StartRetryTask implements Runnable {
+
+    ScheduledExecutorService scheduledExecutorService;
+    ScheduledFuture<?> future;
+
+    private void createTask(long delay) {
+      if (scheduledExecutorService == null) {
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
```

Comment:
Better to use `GrpcUtil.TIMER_SERVICE` or `GrpcUtil.getThreadFactory`

Reply:
ok